### PR TITLE
lib/streamaggr: added rate and rate_avg output

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
### Describe Your Changes

Added `rate` and `rate_avg` output
Resource usage is the same as for increase output, tested on a benchmark

### Checklist

The following checks are **mandatory**:

- [ ] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

File: docs/stream-aggregation.md [hunk 1]
```diff
@@ -577,29 +579,34 @@ The results of `increase` is equal to the following [MetricsQL](https://docs.vic
...
-This is similar to `rate(some_counter[5m])`.
-
 Aggregating irregular and sporadic metrics (received from [Lambdas](https://aws.amazon.com/lambda/)
 or [Cloud Functions](https://cloud.google.com/functions)) can be controlled via [staleness_interval](#staleness) option.
 
 See also [increase_prometheus](#increase_prometheus) and [total](#total).
 
+### rate
+
+`rate` returns the increase of input [time series](https://docs.victoriametrics.com/keyconcepts/#time-series) over the given `interval`.
```
Comment (hagen1778) [inline docs/stream-aggregation.md:597]:
```suggestion
`rate` returns the average per-second increase of input [time series](https://docs.victoriametrics.com/keyconcepts/#time-series) over the given `interval`.
```

File: docs/stream-aggregation.md [hunk 2]
```diff
@@ -577,29 +579,34 @@ The results of `increase` is equal to the following [MetricsQL](https://docs.vic
...
+`rate` makes sense only for aggregating [counters](https://docs.victoriametrics.com/keyconcepts/#counter).
+
+The results of `rate` is equal to the following [MetricsQL](https://docs.victoriametrics.com/metricsql/) query:
```
Comment (hagen1778) [inline docs/stream-aggregation.md:600]:
```suggestion
The results of `rate` are equal to the following [MetricsQL](https://docs.victoriametrics.com/metricsql/) query:
```

File: docs/stream-aggregation.md [hunk 3]
```diff
@@ -577,29 +579,34 @@ The results of `increase` is equal to the following [MetricsQL](https://docs.vic
...
+
+```metricsql
+sum(rate(some_counter[interval]))
+```
+
+`rate` assumes that all the counters start from 0. For example, if the first seen sample for new [time series](https://docs.victoriametrics.com/keyconcepts/#time-series)
```
Comment (hagen1778) [inline docs/stream-aggregation.md:606]: From my reading of https://github.com/VictoriaMetrics/VictoriaMetrics/blob/56531abd562d9a4a4e8825d1a29d9e98c18ec7b8/app/vmselect/promql/rollup.go#L1900-L1935 the `rate` implementation doesn't assume that counter starts from 0. Is my understanding correct?

Comment (AndrewChubatiuk) [reply docs/stream-aggregation.md]: you're right, updated it

Comment (AndrewChubatiuk) [reply docs/stream-aggregation.md]: in this case makes sense to increase a precision of intervals or modify flushState interface. prefer first option

File: docs/stream-aggregation.md [hunk 4]
```diff
@@ -513,6 +513,8 @@ Below are aggregation functions that can be put in the `outputs` list at [stream
 * [count_series](#count_series)
 * [increase](#increase)
 * [increase_prometheus](#increase_prometheus)
+* [rate](#rate)
+* [rate_prometheus](#rate_prometheus)
```
Comment (rparvizian-rbx) [inline docs/stream-aggregation.md:517]: is this implemented?

Comment (hagen1778) [reply docs/stream-aggregation.md]: Do you refer to metricsql function? If yes, then answer is no: such function doesn't exist. The aggregation output won't exist as well.

**Comment by AndrewChubatiuk** [comment]
@hagen1778 tested on a sandbox, resource usage for rate is almost the same as for increase

**Review by hagen1778 — APPROVED** [review]
LGTM

**Comment by hagen1778** [comment]
@AndrewChubatiuk would you mind checking https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6243/commits/8b1033054f4d088a35add60336afb73e97bf55dc ? If you agree with changes then we can merge this pr.

**Comment by valyala** [comment]
@AndrewChubatiuk , @hagen1778 , please see the follow-up commit 5354374b628175f9731802a1086239bb21188651

**Pull request merged**